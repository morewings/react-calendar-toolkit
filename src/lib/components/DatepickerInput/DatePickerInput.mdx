---
name: Datepicker attached to the input
route: '/datepicker-input'
---
import { Playground, Props } from 'docz';
import locale from 'date-fns/esm/locale/ru';
import {DatepickerInput} from 'lib';
import DatepickerInputWithProps from './DatepickerInput';

## Description

`DatepickerInput` component renders `Datepicker`, attached to the provided `input` (DOM node or React component). It contains:
- [`Datepicker`](/datepicker) component without the header.
- [`Input`](/src-lib-components-visual-input-input) which is displayed to user.
- [`PopoverWrapper`](/src-lib-components-visual-popover-popover-wrapper) to display `Datepicker` when user clicks (or makes any other action) inside `Input`, when `mode` prop is set to `popover`.
- [`ModalWrapper`](/src-lib-components-visual-modal-modal-wrapper) to display `Datepicker` when user clicks (or makes any other action) inside `Input`, when `mode` prop is set to `modal`.

<Playground>
  <div>
    <div>Click inside input</div>
    <DatepickerInput onDateSet={date => {console.log('date set', date);}}/>
  </div>
</Playground>

## Props

<Props of={DatepickerInputWithProps} />

## Display mode

### `popover`

Position of popover is calculated dynamically to fit inside the viewport.

<Playground>
  <div>
    <div>Click inside input</div>
    <DatepickerInput
      mode="popover"
      onDateSet={date => {console.log('date set', date);}}
    />
  </div>
</Playground>

### `modal`

Renders `Datepicker` as a modal window. More convenient for a small screen devices.

<Playground>
  <div>
    <div>Click inside input</div>
    <DatepickerInput
      mode="modal"
      onDateSet={date => {console.log('date set', date);}}
    />
  </div>
</Playground>

## Set `Datepicker` props

Set attached [`Datepicker` props](/datepicker#props) via `datePickerProps` property. Same validation rules apply as for `Datepicker` itself, except `onDateSet` callback, `theme` and `dateFnsLocale`, which should be defined in the upper scope.

<Playground>
  {() => {
    const theme = {
      '--bgColor': '#fffff0',
      '--headerBgColor': 'purple',
      '--entrySelectedColor': 'purple',
    };
    return (
      <div>
        <div>Click inside input</div>
        <DatepickerInput
          datePickerProps={{'minPrecision': 'month'}}
          dateFnsLocale={locale}
          theme={theme}
          onDateSet={date => {console.log('date set', date);}}
        />
      </div>
    )
  }}
</Playground>


### `hideOnSelect`

Defines `Datepicker` behavior after user clicks on the date entry.

<Playground>
  <div>
    <div>Click inside input</div>
    <DatepickerInput
      hideOnSelect={false}
      onDateSet={date => {console.log('date set', date);}}
    />
  </div>
</Playground>

### `formatPattern`

Use `date-fns` [format strings](https://date-fns.org/docs/format) to define, what is displayed inside `Input`.

<Playground>
  {() => {
    const theme = {
      '--inputWidth': '160px'
    };
    return (
      <div>
        <div>Click inside input</div>
        <DatepickerInput
          theme={theme}
          formatPattern='GGGG yo Do'
          onDateSet={date => {console.log('date set', date);}}
        />
      </div>
    )
  }}
</Playground>
