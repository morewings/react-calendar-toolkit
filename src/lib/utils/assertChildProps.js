import React, {Fragment} from 'react';

export const trackProps = jest.fn();

export const DummyComponent = props => {
  trackProps(props);
  return <Fragment>Child</Fragment>;
};
