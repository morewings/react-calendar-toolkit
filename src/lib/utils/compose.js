const compose = (...fns) => (...args) =>
  fns.reduceRight(
    (params, f) => (Array.isArray(params) ? f(...params) : f(params)),
    args
  );

export default compose;
