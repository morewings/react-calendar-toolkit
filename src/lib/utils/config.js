export default Object.freeze({
  supportedPrecisions: Object.freeze([
    'year',
    'month',
    'day',
    'hour',
    'minute',
  ]),
});
