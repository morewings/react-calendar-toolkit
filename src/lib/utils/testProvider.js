import React from 'react';
import PropTypes from 'prop-types';
import {Provider as LocaleProvider} from 'lib/features/locale/withLocaleContext';
import {Provider as ThemeProvider} from 'lib/features/theme/withTheme';
import {Provider as ModalProvider} from 'lib/features/modal/withModalProvider';
import {Provider as DatepickerProvider} from 'lib/features/datepicker/withDatepickerProvider';

export const DatepickerMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <DatepickerProvider>{props.children}</DatepickerProvider>
    </ThemeProvider>
  </LocaleProvider>
);

export const InputMockProvider = props => (
  <LocaleProvider {...props}>
    <ThemeProvider>
      <ModalProvider>
        <DatepickerProvider>{props.children}</DatepickerProvider>
      </ModalProvider>
    </ThemeProvider>
  </LocaleProvider>
);

DatepickerMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

InputMockProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
