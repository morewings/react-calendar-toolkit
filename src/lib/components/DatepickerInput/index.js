import DatepickerInput from './DatepickerInput';

export default DatepickerInput;
