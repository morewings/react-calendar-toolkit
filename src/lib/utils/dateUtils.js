import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
