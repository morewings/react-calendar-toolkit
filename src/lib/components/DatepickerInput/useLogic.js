import {useCallback} from 'react';
import {useModalActions} from 'lib/features/modal';
import {convertToDate} from 'lib/utils/dateUtils';
import {useFormatDate} from 'lib/features/locale';
import {useDatepickerContext} from 'lib/features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatepicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatepickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatepicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatepicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
