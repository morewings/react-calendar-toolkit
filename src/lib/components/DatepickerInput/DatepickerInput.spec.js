import React from 'react';
import {render} from '@testing-library/react';
import {InputMockProvider as Provider} from 'lib/utils/testProvider';
import {assertChildProps} from 'lib/utils/assertChildProps';
import DatepickerInput from './DatepickerInput';

const [Datepicker, DatepickerProps] = assertChildProps();
const [Input, InputProps] = assertChildProps();
const [PopoverProvider, PopoverProviderProps] = assertChildProps();
const [PopoverWrapper, PopoverWrapperProps] = assertChildProps();
const [ModalProvider, ModalProviderProps] = assertChildProps();
const [ModalWrapper, ModalWrapperProps] = assertChildProps();

const onDateSet = jest.fn();

const renderWithProps = (props = {}) =>
  render(
    <DatepickerInput
      renderDatepickerAs={Datepicker}
      renderInputAs={Input}
      popoverProvider={PopoverProvider}
      wrapPopoverWith={PopoverWrapper}
      modalProvider={ModalProvider}
      wrapmodalWith={ModalWrapper}
      {...props}
      today={new Date(2020, 0, 10)}
      onDateSet={onDateSet}
    />,
    {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    }
  );

describe('Datepicker', () => {
  beforeEach(() => {
    onDateSet.mockClear();
    DatepickerProps.mockClear();
    InputProps.mockClear();
    PopoverProviderProps.mockClear();
    PopoverWrapperProps.mockClear();
    ModalProviderProps.mockClear();
    ModalWrapperProps.mockClear();
  });

  it('renders with Popover by default', () => {
    const {asFragment} = renderWithProps();
    expect(asFragment()).toMatchSnapshot();
    expect(PopoverProviderProps.mock.calls[0][0]).toMatchSnapshot();
  });

  it('renders with Modal if mode === `modal`', () => {
    const {asFragment} = renderWithProps({mode: 'modal'});
    expect(asFragment()).toMatchSnapshot();
    expect(ModalProviderProps.mock.calls[0][0]).toMatchSnapshot();
  });
});
