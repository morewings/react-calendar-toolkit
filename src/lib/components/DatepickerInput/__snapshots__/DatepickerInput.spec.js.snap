// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker renders with Modal if mode === \`modal\` 1`] = `
<DocumentFragment>
  <div
    data-testid="Modal"
  >
    <div>
      <div>
        Input
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Datepicker renders with Modal if mode === \`modal\` 2`] = `
Object {
  "children": <ForwardRef>
    <mockConstructor
      date={2020-01-08T00:00:00.000Z}
      onChange={[MockFunction]}
      toggleDatepicker={[Function]}
      value="01/08/2020"
    />
  </ForwardRef>,
  "isVisible": false,
  "renderDatepickerAs": [Function],
  "toggleDatepicker": [Function],
  "wrapModalWith": [Function],
  "wrapPopoverWith": [MockFunction],
}
`;

exports[`Datepicker renders with Modal if mode === \`modal\` 3`] = `
Object {
  "date": 2020-01-08T00:00:00.000Z,
  "onChange": [MockFunction],
  "toggleDatepicker": [Function],
  "value": "01/08/2020",
}
`;

exports[`Datepicker renders with Popover by default 1`] = `
<DocumentFragment>
  <div
    data-testid="Popover"
  >
    <div>
      <div>
        Input
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Datepicker renders with Popover by default 2`] = `
Object {
  "children": <ForwardRef>
    <mockConstructor
      date={2020-01-08T00:00:00.000Z}
      onChange={[MockFunction]}
      toggleDatepicker={[Function]}
      value="01/08/2020"
    />
  </ForwardRef>,
  "isVisible": false,
  "renderDatepickerAs": [Function],
  "toggleDatepicker": [Function],
  "wrapModalWith": [Function],
  "wrapPopoverWith": [MockFunction],
}
`;

exports[`Datepicker renders with Popover by default 3`] = `
Object {
  "date": 2020-01-08T00:00:00.000Z,
  "onChange": [MockFunction],
  "toggleDatepicker": [Function],
  "value": "01/08/2020",
}
`;
